connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -866,4 +867,18 @@ public List<ConfigKeyInfo> connectorPluginConfig(String pluginName) {
         }
     }
 
+    @Override
+    public void connectorOffsets(String connName, Callback<ConnectorOffsets> cb) {
```

Comment:
I've tweaked the log statement and used a snapshot throughout for the consistent view in the `AbstractHerder` method but I'm still not sure I understand why we need to put this through the tick thread in the `DistributedHerder`. I agree that there shouldn't be concerns about unnecessarily blocking that thread but since the slight chance of concurrent connector deletion exists in both scenarios, why would we want to do so for a request that is non-mutating?

Reply:
Gotcha, minimising / avoiding concurrent calls to the `Worker` does seem like a worthwhile property to preserve. I've overridden the `AbstractHerder::connectorOffsets` method in `DistributedHerder` to put it through the tick thread and in `StandaloneHerder` to make it a synchronized method. I haven't added an additional call to `ConfigBackingStore::refresh` since we're capturing a fresh snapshot (in the `AbstractHerder`) and not using an existing stale one so we should be reasonably up to speed on the state of the topic (due to the work thread in the underlying Kafka based log). Btw on a side note, we don't seem to be making calls to `ConfigBackingStore::refresh` even in methods like `putConnectorConfig` where it does seem like it would be useful to do so, what do you think?

Reply:
> I haven't added an additional call to `ConfigBackingStore::refresh` since we're capturing a fresh snapshot (in the `AbstractHerder`) and not using an existing stale one so we should be reasonably up to speed on the state of the topic (due to the work thread in the underlying Kafka based log)

I won't push for this but it is worth pointing out that the worker may have lost contact with the config topic at some point and so we can't assume that taking a snapshot gives us an up-to-date view of it. I do agree that no matter what, the risks here are small and the guarantees we're providing already should be fine for now.

> Btw on a side note, we don't seem to be making calls to `ConfigBackingStore::refresh` even in methods like `putConnectorConfig` where it does seem like it would be useful to do so, what do you think?

The `putConnectorConfig` example is an interesting one. The leader will force a read-to-end of the config topic after writing connector configs or tombstones (that delete connector configs) to it, so it's actually guaranteed that the leader will be up-to-date on any newly-created or just-destroyed connectors. I haven't done a comprehensive analysis of all of the herder methods but I suspect this kind of logic holds for most of them. The only exceptions I can think of off the top of my head are pausing/resuming a connector, but that's low-stakes enough that a read-to-end probably isn't worth it.


I'm going to resolve this conversation since I'm satisfied with the way things look now. I'm happy to look at a follow-up PR or ticket if there's room for improvement with the herder that you'd like to tackle.